gluten-core/src/main/scala/io/glutenproject/execution/BatchScanExecTransformer.scala

```diff
@@ -130,4 +131,16 @@ class BatchScanExecTransformer(
     case "ClickHouseScan" => ReadFileFormat.MergeTreeReadFormat
     case _ => ReadFileFormat.UnknownFormat
   }
+
+  override def doCanonicalize(): BatchScanExecTransformer = {
+    val canonicalized = super.doCanonicalize()
+    new BatchScanExecTransformer(
+      canonicalized.output,
+      canonicalized.scan,
+      canonicalized.runtimeFilters,
+      QueryPlan.normalizePredicates(
+        pushdownFilters.filterNot(_ == DynamicPruningExpression(Literal.TrueLiteral)),
```

Comment:
not related to this pr, but it seems `pushdownFilters` is never been set and used... not sure the original idea, maybe we can just remove `pushdownFilters`

Reply:
Yes, I also noticed this when support iceberg read. @PHILO-HE @rui-mo  Can we remove it?

Reply:
Yes, we can remove it in this PR. Thanks!

Reply:
@ulysses-you @PHILO-HE I encountered an error during the test. This parameter here cannot be deleted. Although it is not used, it is possible that the parameter was added unintentionally that avoid the error. 
```
Exception in thread "main" java.lang.IllegalStateException:
Failed to copy node.
Is otherCopyArgs specified correctly for BatchScanExecTransformer.
Exception message: wrong number of arguments
ctor: public io.glutenproject.execution.BatchScanExecTransformer(scala.collection.Seq,org.apache.spark.sql.connector.read.Scan,scala.collection.Seq)?
types: class scala.collection.immutable.$colon$colon, class org.apache.iceberg.spark.source.SparkBatchQueryScan, class scala.collection.immutable.$colon$colon, class scala.None$
args: List(application_id#10037, executor_id#10046, start_time#10047L, end_time#10048L, dt#10079, emr_cluster_id#10080, emr_user_id#10081), IcebergScan(table=default_iceberg.doctor_oss.spark_task_info, type=struct<1: application_id: optional string, 10: executor_id: optional string, 11: start_time: optional long, 12: end_time: optional long, 43: dt: optional string, 44: emr_cluster_id: optional string, 45: emr_user_id: optional string>, filters=[not_null(ref(name="dt")), not_null(ref(name="emr_user_id")), ref(name="dt") == "2023-08-01", not(ref(name="emr_user_id") == "1836697956323238"), not_null(ref(name="emr_cluster_id")), not_null(ref(name="application_id"))], runtimeFilters=[], caseSensitive=false), List(dynamicpruningexpression(true), dynamicpruningexpression(emr_cluster_id#10080 IN dynamicpruning#10622)), None

	at org.apache.spark.sql.catalyst.trees.TreeNode.makeCopy(TreeNode.scala:854)
	at org.apache.spark.sql.catalyst.trees.TreeNode.makeCopy(TreeNode.scala:797)
	at org.apache.spark.sql.execution.SparkPlan.super$makeCopy(SparkPlan.scala:100)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$makeCopy$1(SparkPlan.scala:100)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.execution.SparkPlan.makeCopy(SparkPlan.scala:100)
	at org.apache.spark.sql.execution.SparkPlan.makeCopy(SparkPlan.scala:60)
	at org.apache.spark.sql.catalyst.plans.QueryPlan.mapExpressions(QueryPlan.scala:223)
	at org.apache.spark.sql.catalyst.plans.QueryPlan.transformExpressionsUpWithPruning(QueryPlan.scala:188)
	at org.apache.spark.sql.execution.reuse.ReuseExchangeAndSubquery$$anonfun$org$apache$spark$sql$execution$reuse$ReuseExchangeAndSubquery$$reuse$1$1.applyOrElse(ReuseExchangeAndSubquery.scala:54)
	at org.apache.spark.sql.execution.reuse.ReuseExchangeAndSubquery$$anonfun$org$apache$spark$sql$execution$reuse$ReuseExchangeAndSubquery$$reuse$1$1.applyOrElse(ReuseExchangeAndSubquery.scala:44)
```

It may be a little difficult to explain here. I try to make it clear. When Spark is in TreeNode.makeCopy, it will call the productArity method of Product to obtain the number of parameters of the case class. The current number of parameters of BatchScanExecTransformer is 3, but productArity gets 4. After my test, productArity gets the number of parameters of the first specific case class, which is the number of parameters of BatchScanExec, which does not match the number of BatchScanExecTransformer constructors. 

I think the standard approach here is that if we need to inherit Spark's case class plan, we need to ensure that the number of parameters is consistent, otherwise it will cause problems when plan.transform calls makeCopy. Or implement the otherCopyArgs method to ensure that not transformed args can be copied.

Reply:
It's a subtle issue. Why `productArity` returns 4 for `BatchScanExec`? It looks `BatchScanExec` has only 3 constructor args.

Reply:
> It's a subtle issue. Why `productArity` returns 4 for `BatchScanExec`? It looks `BatchScanExec` has only 3 constructor args.

Spark 3.3 has 4 args.
https://github.com/apache/spark/blob/b170a670b68616af49b26cbcc84894c939713837/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala#L40

Reply:
```scala
case class A(a1: Int, a2: Int) extends Product
class B(a1: Int, a2: Int, a3: Int) extends A(a1, a2)
val b = new B(1, 2, 3)
b.productArity
```
b.productArity will return 2.

Reply:
I see. Please help fix it in a separate PR and leave some comments to explain in source code. Thanks!